cedar-policy-core/src/entities/json/schema_types.rs

```diff
@@ -382,10 +378,10 @@ pub fn schematype_of_restricted_expr(
                         BorrowedRestrictedExpr::new_unchecked(v), // assuming the invariant holds for the record as a whole, it will also hold for each attribute value
                         extensions,
                     )?;
-                    // we can't know if the attribute is required or optional,
-                    // but marking it optional is more flexible -- allows the
-                    // attribute type to `is_consistent_with()` more types
-                    Ok((k.clone(), AttributeType::optional(attr_type)))
+                    // We can't know if the attribute is required or optional.
+                    // Assume it's required so that calling `is_consistent_with` on this and a type from the Schema
+                    // will reject attributes not present in the schema.
+                    Ok((k.clone(), AttributeType::required(attr_type)))
```

Comment:
I think this is the correct choice. `required <: optional`, and we know the attribute is present, so there's no reason to leave it as optional.

Reply:
Actually this doesn't work because it marks more sets as heterogeneous. I'll add a comment in the issues.

Reply:
I don't see why that would be the case. Do you mean `[{a: 1}, {b: 2}]` is treated as heterogeneous when the types are `{a: Long}` and `{b: Long}`, but not when it's `{a?: Long}` and `{b?: Long}`? That doesn't make sense to me.